axelrod/player.py

```diff
@@ -122,6 +125,44 @@ def __init__(self):
         self.state_distribution = defaultdict(int)
         self.set_match_attributes()
 
+
+    def __eq__(self, other):
+        """
+        Test if two players are equal.
+        """
+        if self.__repr__() != other.__repr__():
+            return False
+        for attribute in set(list(self.__dict__.keys()) +
+                             list(other.__dict__.keys())):
+
+            value = getattr(self, attribute, None)
+            other_value = getattr(other, attribute, None)
+
+            if isinstance(value, np.ndarray):
+                if not (np.array_equal(value, other_value)):
+                    return False
+
+            elif isinstance(value, types.GeneratorType) or \
+                 isinstance(value, itertools.cycle):
+
+                # Split the original generator so it is not touched
+                generator, original_value = itertools.tee(value)
+                other_generator, original_other_value = itertools.tee(other_value)
+
+                setattr(self, attribute,
+                        (ele for ele in original_value))
+                setattr(other, attribute,
+                        (ele for ele in original_other_value))
+
+                if not (all(next(generator) == next(other_generator)
```

Comment:
Yup that's right. What `itertools.tee` does is essentially fork a generator in to two versions of itself from the state it currently is in. 

That is also being tests:

1. Create two `Cooperators` and give them two cycle generators and tests that they are equal (at this stage their generator are in the same state:

    ```
    205         p1 = axelrod.Cooperator()                                                                                                                 
    206         p2 = axelrod.Cooperator()                                                                                                                 
    207         # Check that the cycle is still the same                                                                                                
    208         p1.cycle = itertools.cycle(range(10))                                                                                                    
    209         p2.cycle = itertools.cycle(range(10))                                                                                                    
    210         self.assertEqual(p1, p2)   
    ```

2. Bump `p2`s cycle and now the players are not the same:

    ```
    212         _ = next(p2.cycle)                                                                                                                       
    213         self.assertNotEqual(p1, p2)  
    ```

3. Check what the next values in the cycle are for each player, for `p1` it should be `0` because the player has not iterated through any values and for `p2` it should be `1` because it has iterated through the first value:


    ```
    215         # Check that internal generator object has not been changed                                                                              
    216         self.assertEqual(next(p1.cycle), 0)                                                                                                       
    217         self.assertEqual(next(p2.cycle), 1)  
    ```

Let me know if you think I could clarify that in the code and/or add other tests.